File: istaroth/agd/renderables/material.py
"""Material processing and rendering."""

import hashlib

from istaroth import utils
from istaroth.agd import (
    id_types,
    issues,
    processed_types,
    repo,
    text_utils,
)
from istaroth.text import types as text_types


def get_material_info(
    material_id: id_types.MaterialId, *, data_repo: repo.DataRepo
) -> processed_types.MaterialInfo:
    """Get material information for a specific material ID."""
    text_map = data_repo.load_text_map()
    material_tracker = data_repo.load_material_excel_config_data()

    material = material_tracker.get(material_id)
    if material is None:
        raise ValueError(f"Material with ID {material_id} not found")

    name_hash = material["nameTextMapHash"]
    if (name := text_map.get_optional(name_hash)) is None:
        issues.record(issues.IssueType.MISSING_MATERIAL_NAME, str(name_hash))
        name = "Unknown Material"

    desc_hash = material["descTextMapHash"]
    if (description := text_map.get_optional(desc_hash)) is None:
        issues.record(issues.IssueType.MISSING_MATERIAL_DESC, str(desc_hash))
        description = "No description available"

    return processed_types.MaterialInfo(
        material_id=material_id, name=name, description=description
    )


def _humanize_material_type(material_type: str) -> str:
    """Turn a raw MaterialType enum into a readable title."""
    return material_type.removeprefix("MATERIAL_").replace("_", " ").title()


def render_materials_by_type(
    material_type: str, materials: list[processed_types.MaterialInfo]
) -> processed_types.RenderedItem:
    """Render multiple materials of the same type into a single RAG-suitable file."""
    material_type_id = int(
        hashlib.sha256(material_type.encode("utf-8")).hexdigest()[:12], base=16
    )

    safe_type = utils.make_safe_filename_part(material_type)
    filename = f"{material_type_id}_{safe_type}.txt"

    material_type_name = _humanize_material_type(material_type)

    content_lines = [f"# Materials: {material_type_name}\n"]

    sorted_materials = sorted(materials, key=lambda x: x.material_id)

    for material_info in sorted_materials:
        content_lines.append(f"## {material_info.name}")
        content_lines.append("")
        content_lines.append(material_info.description)
        content_lines.append("")

    rendered_content = "\n".join(content_lines).rstrip()

    return processed_types.RenderedItem(
        text_metadata=text_types.TextMetadata(
            category=text_types.TextCategory.AGD_MATERIAL_TYPE,
            title=material_type_name,
            id=material_type_id,
            relative_path=f"{text_types.TextCategory.AGD_MATERIAL_TYPE.value}/{filename}",
        ),
        content=rendered_content,
    )

File: rust/istaroth-agd-regen/src/renderables/material.rs
//! Port of renderables/material.py.

use crate::firstseen::Domain;
use crate::issues::{IssueType, Scope};
use crate::rendered_item::RenderedItem;
use crate::repo::Repo;
use crate::util;
use crate::vh::ValueExt;
use anyhow::Result;
use rustc_hash::FxHashSet;

/// MaterialTypes pass discovery: sorted distinct materialType values.
pub fn discover(repo: &Repo) -> Result<Vec<String>> {
    let mut types: FxHashSet<String> = FxHashSet::default();
    for material in repo.excel.material.values() {
        types.insert(material.s("materialType")?.to_string());
    }
    let mut v: Vec<String> = types.into_iter().collect();
    v.sort();
    Ok(v)
}

/// MaterialTypes: all same-type materials into one file.
pub fn process(repo: &Repo, scope: &Scope, material_type: &str) -> Result<Option<RenderedItem>> {
    struct MaterialInfo {
        material_id: i64,
        name: String,
        description: String,
    }
    let mut materials: Vec<MaterialInfo> = Vec::new();
    for material in repo.excel.material.values() {
        if material.s("materialType")? != material_type {
            continue;
        }
        let material_id = material.i("id")?;
        let name_hash = material.i("nameTextMapHash")?;
        let name = match repo.tm.get_optional(name_hash, scope)? {
            Some(name) => name,
            None => {
                scope.record_issue(IssueType::MissingMaterialName, name_hash.to_string());
                "Unknown Material".to_string()
            }
        };
        let desc_hash = material.i("descTextMapHash")?;
        let description = match repo.tm.get_optional(desc_hash, scope)? {
            Some(desc) => desc,
            None => {
                scope.record_issue(IssueType::MissingMaterialDesc, desc_hash.to_string());
                "No description available".to_string()
            }
        };
        if util::should_skip_text(&name, repo.language)
            || util::should_skip_text(&description, repo.language)
        {
            continue;
        }
        materials.push(MaterialInfo {
            material_id,
            name,
            description,
        });
    }
    if materials.is_empty() {
        return Ok(None);
    }

    let material_type_id = util::sha256_id(material_type);
    let safe_type = util::make_safe_filename_part(material_type);
    let material_type_name = util::py_title_case(
        &material_type
            .strip_prefix("MATERIAL_")
            .unwrap_or(material_type)
            .replace('_', " "),
    );
    let mut content_lines = vec![format!("# Materials: {material_type_name}\n")];
    materials.sort_by_key(|m| m.material_id);
    for m in &materials {
        content_lines.push(format!("## {}", m.name));
        content_lines.push(String::new());
        content_lines.push(m.description.clone());
        content_lines.push(String::new());
    }
    let content = util::py_rstrip(&content_lines.join("\n")).to_string();
    let versions = repo
        .first_seen
        .resolve_ints(Domain::Material, materials.iter().map(|m| m.material_id))?;
    Ok(Some(RenderedItem::new(
        "agd_material_type",
        material_type_name,
        material_type_id,
        format!("{material_type_id}_{safe_type}.txt"),
        versions,
        content,
    )))
}
